integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs

```diff
@@ -36,55 +49,111 @@ async fn handle_body(
-        Ok(true) => (),
-        Ok(false) => {
-            if database_creation_enabled()? {
-                thread::sleep(time::Duration::from_secs(TIMESTREAM_API_WAIT_SECONDS));
-                create_database(client, &database_name).await?;
-            } else {
-                return Err(anyhow!(
-                    "Database {} does not exist and database creation is not enabled",
-                    database_name
-                ));
-            }
-        }
-        Err(error) => return Err(anyhow!(error)),
-    }
-
-    for (table_name, _) in records.iter() {
-        match table_exists(client, &database_name, table_name).await {
+    if let Ok(true) = std::env::var("enable_database_creation").map(env_var_to_bool) {
+        match database_exists(client, &database_name).await {
             Ok(true) => (),
             Ok(false) => {
-                if table_creation_enabled()? {
+                if database_creation_enabled()? {
                     thread::sleep(time::Duration::from_secs(TIMESTREAM_API_WAIT_SECONDS));
-                    create_table(client, &database_name, table_name, get_table_config()?).await?
+                    create_database(client, &database_name).await?;
                 } else {
                     return Err(anyhow!(
-                        "Table {} does not exist and database creation is not enabled",
-                        table_name
+                        "Database {} does not exist and database creation is not enabled",
+                        database_name
                     ));
                 }
             }
-            Err(error) => println!("error checking table exists: {:?}", error),
+            Err(error) => return Err(anyhow!(error)),
         }
     }
 
-    for (table_name, mut records) in records.into_iter() {
-        ingest_records(client, &database_name, &table_name, &mut records).await?
+    // Use a semaphore to limit the maximum number of threads used to process batches in parallel
+    let semaphore = Arc::new(Semaphore::new(NUM_BATCH_THREADS));
+    let mut batch_ingestion_futures = FuturesUnordered::new();
+
+    // Track total time taken to check existence of tables and ingest records
+    let ingestion_start = Instant::now();
+
+    // Ingest records for each table, in parallel
+    for (table_name, records) in records {
+        let permit = semaphore
+            .clone()
+            .acquire_owned()
+            .await
+            .expect("Failed to get semaphore permit");
+
+        // Use Arc::clone to create a shallow clone of the client
+        let client_clone = Arc::clone(client);
+        let database_name_clone = Arc::clone(&database_name);
+
+        // Create a future for ingesting to the current table
+        let future = task::spawn(async move {
+            // Check whether the table exists
+            if let Ok(true) = std::env::var("enable_table_creation").map(env_var_to_bool) {
```

Comment:
Let's the logic for table creation to `timestream_utils`. So we can have something more like the following instead:

```rust
if let Ok(true) = std::env::var("enable_table_creation").map(env_var_to_bool) {
    create_table_if_non_existent(&client_clone, &database_name_clone, &table_name);
}
```

Reply:
Okay, I have extracted that section into its own function:
```rust
pub async fn create_table_if_non_existent(
    client: &Arc<timestream_write::Client>,
    database_name: &Arc<String>,
    table_name: &str,
) -> Result<(), Error> {
    match table_exists(client, database_name, table_name).await {
        Ok(true) => (),
        Ok(false) => {
            thread::sleep(time::Duration::from_secs(TIMESTREAM_API_WAIT_SECONDS));
            create_table(client, database_name, table_name, get_table_config()?).await?
        }
        Err(error) => info!("error checking table exists: {:?}", error),
    }

    Ok(())
}
```